include/openssl/sha.h

```diff
@@ -82,6 +88,13 @@ OSSL_DEPRECATEDIN_3_0 unsigned char *SHA256(const unsigned char *d, size_t n,
                                             unsigned char *md);
 OSSL_DEPRECATEDIN_3_0 void SHA256_Transform(SHA256_CTX *c,
                                             const unsigned char *data);
+# else
+#  ifdef OPENSSL_COMPENSATE_DEPRECATION
+#   define SHA224(d, n, md) \
+    (EVP_Digest(d, n, md, NULL, EVP_sha224(), NULL) ? md : NULL)
+#   define SHA256(d, n, md) \
+    (EVP_Digest(d, n, md, NULL, EVP_sha256(), NULL) ? md : NULL)
```

Comment:
`EVP_sha256()` should be a string `"SHA256"` or `SN_sha256` and the algorithm fetched in `EVP_Digest()`.

Similarly everywhere else :grinning:

Reply:
Hmm, `EVP_Digest()` was already there!
Can't we simply keep and use it as-is?
Of course, I can add `EVP_Q_digest()` which takes a string instead, plus a libctx and query string:
```
int EVP_Q_digest(const void *data, size_t count,
                 unsigned char *md, unsigned int *size,
                 OSSL_LIB_CTX *libctx, const char *name, const char *propq);
```

Reply:
I think adding the EVP_Q_digest with the parameters as you propose is a good idea.

Reply:
Done.

Reply:
Sorry for bringing up this thread, but what's the problem with avoiding `EVP_sha256()`? And generally why would `EVP_Q_Digest()` be preferred over `EVP_Digest()`, since the former wraps around the latter?

Asking this from an API user's perspective since I'm not much familiar with ossl internals

Reply:
I'm not an expert on the internals myself, so can just answer from user' perspective.
As far as I understand, the idea behind all those API changes was that algorithm names should no more be hard-coded in the function/macro names, in order to make the API fully crypto agile.
Though I think this would have been possible also with `EVP_Digest()`.

Reply:
The EVP_sha256() uses so called implicit fetch of the algorithm. In the context of EVP_Digest vs EVP_Q_Digest it does not mean much apart from the fact, that EVP_Q_Digest can be used with any provided hash even such that OpenSSL libcrypto itself does not know anything about. It just needs to be available within some loaded third party provider.